modules/web3/events.ts

```diff
@@ -0,0 +1,120 @@
+    }
+    const range = toBlock - fromBlock;
+
+    return Promise.all(
+        // Getting logs in batches of max blocks allowed by RPC
+        Array.from({ length: Math.ceil(range / rpcMaxBlockRange) }, (_, i) => i).map(async (i) => {
+            const from = fromBlock + (i > 0 ? 1 : 0) + i * rpcMaxBlockRange;
+            const to = Math.min(fromBlock + (i + 1) * rpcMaxBlockRange, toBlock);
+
+            // Usually RPCs are handling any number of addresses, but it here batching just to be on the safe side
+            const logRequests: Promise<Event[]>[] = chunk(addresses, 500).map(async (addresses) => {
+                // Fetch logs with a raw json request until we support Viem or Ethers6
+                const payload = {
+                    jsonrpc: '2.0',
+                    id: 1,
+                    method: 'eth_getLogs',
+                    params: [
+                        {
+                            address: addresses,
+                            topics: topics.length === 1 ? topics : [topics],
+                            fromBlock: '0x' + BigInt(from).toString(16),
+                            toBlock: '0x' + BigInt(to).toString(16),
+                        },
+                    ],
+                };
+
+                return fetch(rpcUrl, {
+                    method: 'POST',
+                    headers: {
+                        'Content-Type': 'application/json',
+                    },
+                    body: JSON.stringify(payload),
+                })
+                    .then(
+                        (response) =>
+                            response.json() as Promise<
+                                { result: Event[] } | { error: { code: string; message: string } }
+                            >,
+                    )
+                    .then((response) => {
+                        if ('error' in response) {
+                            return Promise.reject(response.error.message);
+                        }
+
+                        return response.result;
+                    })
+                    .catch((e: any) => {
+                        // Ankr RPC returns error if block range is too wide
+                        if (e.includes('block range is too wide')) {
+                            return getEvents(from, to, addresses, topics, rpcUrl, rpcMaxBlockRange / 2);
+                        }
+
+                        // Infura returns 'more than 10000 results' error if block range is too wide
+                        if (e.includes('query returned more than 10000 results')) {
+                            const range = e
+                                .match(/\[([0-9a-fA-F, x]+)\]/)
+                                .pop()
+                                .split(', ')
+                                .map((hex: string) => parseInt(hex, 16));
+
+                            return getEvents(from, to, addresses, topics, rpcUrl, range[1] - range[0]);
+                        }
```

Comment:
These retries will mess with the "latestSyncBlock" logic. If it retries, it will still persist the toBlock as the latest think block and not the one it actually synced to. You either need to make sure all blocks up until toBlock are queried or get rid of the retries and set reasonable block limits.

I guess we anyway need to make sure that it fails if one of the promises fails, or not do promise.all but do they sequentially and return the block number until which we synced

Reply:
thanks, I double checked and unless I'm missing something obvious, catch is refetching only the chunk that failed, not the whole range, so it's just filling the gaps. there is promise.all instead of allSettled to make sure it gets rejected as a whole, not just a part of it.

calling getEvents should return the whole range or nothing.